babel/src/test/java/org/apache/calcite/test/BabelParserTest.java

```diff
@@ -1667,4 +1667,52 @@ private void checkParseInfixCast(String sqlType) {
         + "FROM `BAR`";
     sql(sql).ok(expected);
   }
+
+  @Test void testTopNNoPercentNoTies() {
+    final String sql = "select top 5 bar from foo";
+    final String expected = "SELECT TOP 5 `BAR`\n"
+        + "FROM `FOO`";
+    sql(sql).ok(expected);
+  }
+
+  @Test void testTopNWithTies() {
+    final String sql = "select top 5 with ties bar from foo";
+    final String expected = "SELECT TOP 5 WITH TIES `BAR`\n"
+        + "FROM `FOO`";
+    sql(sql).ok(expected);
+  }
+
+  @Test void testTopNWithPercent() {
+    final String sql = "select top 5.2 percent bar from foo";
+    final String expected = "SELECT TOP 5.2 PERCENT `BAR`\n"
+        + "FROM `FOO`";
+    sql(sql).ok(expected);
+  }
+
+  @Test void testTopNWithPercentWithTies() {
+    final String sql = "select top 5 percent with ties bar from foo";
+    final String expected = "SELECT TOP 5 PERCENT WITH TIES `BAR`\n"
+        + "FROM `FOO`";
+    sql(sql).ok(expected);
+  }
+
+  @Test void testTopNGreaterThan100WithoutPercent() {
+    final String sql = "select top 500 bar from foo";
+    final String expected = "SELECT TOP 500 `BAR`\n"
+        + "FROM `FOO`";
+    sql(sql).ok(expected);
+  }
+
+  @Test void testTopNDecimalWithoutPercentFails() {
+    final String sql = "select top ^5.2^ bar from foo";
+    final String expected = "(?s).*Cannot specify non-integer value."
+        + "*without specifying PERCENT.*";
+    sql(sql).fails(expected);
+  }
+
+  @Test void testTopNPercentGreaterThan100Fails() {
+    final String sql = "select top 200 ^percent^ bar from foo";
+    final String expected = "(?s).*Numeric literal.*out of range.*";
+    sql(sql).fails(expected);
+  }
```

Comment:
Could add a test here for negative numbers to show that it should fail on that input

Reply:
Good call, will do.